src/dtree.ml

```diff
@@ -0,0 +1,648 @@
+      match elt with
+      | Symb(_, _), _
+      | Vari(_)   , _ -> [| |]
+      | Appl(_, _), p ->
+         (* ^ Arguments verified in filter *)
+         let upat = fst @@ Basics.get_args pat in
+         let hs, hargs = Basics.get_args (fst elt) in
+         let np = Sub.sub p in
+         let tagged = Sub.tag ~empty:np (Array.of_list hargs) in
+         Array.append (spec_transform upat (hs, np)) tagged
+      | _             -> (* Cases that require the pattern *)
+      match elt, pat with
+      | (Patt(_, _, e), p), Appl(_, _) ->
+         let arity = List.length @@ snd @@ Basics.get_args pat in
+         let tagged = Sub.tag
+           (Array.init arity (fun _ -> Patt(None, "", e))) in
+         (Array.map (fun (te, po) -> (te, Sub.prefix p po)) tagged)
+      | (Patt(_, _, _), _)    , _      -> [| |]
+      | _                              -> assert false
+
+(** [specialize p c m] specializes the matrix [m] when matching pattern [p]
+    against column [c].  A matrix can be specialized by a user defined symbol.
+    In case an {!constructor:Appl} is given as pattern [p], only terms having
+    the same number of arguments and the same leftmost {e non}
+    {!constructor:Appl} term match. *)
+  let specialize : term -> int -> rule list -> rule list = fun p ci rs ->
+    let filtered = List.filter (fun { lhs ; _} ->
+      spec_filter p (fst lhs.(ci))) rs in
+    let newcith = List.map (fun { lhs ; _ } -> spec_transform p lhs.(ci))
+      filtered in
+    List.map2 (fun newcith rul ->
+      let { lhs ; _ } = rul in
+      let postfix = if ci >= Array.length lhs then [| |] else
+          Array.sub lhs (ci + 1) (Array.length lhs - (ci + 1)) in
+      let newline = Array.concat
+        [ Array.sub lhs 0 ci
+        ; newcith
+        ; postfix ] in
+      { rul with lhs = newline }) newcith filtered
+
+  (** [default c m] computes the default matrix containing what remains to be
+      matched in case the pattern used is not in the column [c]. *)
+  let default : int -> rule list -> rule list = fun ci rs ->
+    let filtered = List.filter (fun { lhs ; _ } ->
+      match fst @@ lhs.(ci) with
+      | Patt(_ , _, _) -> true
+      | Symb(_, _)
+      | Abst(_, _)
+      | Vari(_)
+      | Appl(_, _)     -> false
+      | _              -> assert false) rs in
+    List.map (fun rul ->
+      let { lhs ; _ } = rul in
+      match lhs.(ci) with
+      | Patt(_, _, _), _ ->
+         let postfix = if (ci + 1) >= Array.length lhs then [| |] else
+             Array.sub lhs (ci + 1) (Array.length lhs - (ci + 1)) in
+         { rul with lhs = Array.append (Array.sub lhs 0 ci) postfix  }
+      | _                -> assert false) filtered
+end
+
+module Cm = ClauseMat
+
+(** {b Note} The compiling step creates a tree ready to be used for pattern
+    matching.  A tree guides the pattern matching by
+    + accepting constructors and filtering possible rules,
+    + getting the most efficient term to match in the stack,
```

Comment:
What does that mean "most efficient term"? Do you mean "best argument"?

Reply:
In a way, best in the sense of depth of the compiled tree, although I'm not sure that "best argument" is more explicit than "most efficient".